src/main/java/org/junitpioneer/jupiter/resource/ResourceExtension.java

```diff
@@ -0,0 +1,368 @@
+ *
+ * http://www.eclipse.org/legal/epl-v20.html
+ */
+
+package org.junitpioneer.jupiter.resource;
+
+import static java.util.Arrays.asList;
+import static java.util.Collections.unmodifiableList;
+import static java.util.Comparator.comparing;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.toList;
+
+import java.lang.reflect.Constructor;
+import java.lang.reflect.Executable;
+import java.lang.reflect.Method;
+import java.lang.reflect.Parameter;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Stream;
+
+import org.junit.jupiter.api.extension.DynamicTestInvocationContext;
+import org.junit.jupiter.api.extension.ExtensionContext;
+import org.junit.jupiter.api.extension.InvocationInterceptor;
+import org.junit.jupiter.api.extension.ParameterContext;
+import org.junit.jupiter.api.extension.ParameterResolutionException;
+import org.junit.jupiter.api.extension.ParameterResolver;
+import org.junit.jupiter.api.extension.ReflectiveInvocationContext;
+import org.junit.platform.commons.support.AnnotationSupport;
+import org.junit.platform.commons.support.ReflectionSupport;
+
+class ResourceExtension implements ParameterResolver, InvocationInterceptor {
+
+	private static final ExtensionContext.Namespace NAMESPACE = //
+		ExtensionContext.Namespace.create(ResourceExtension.class);
+
+	private static final AtomicLong KEY_GENERATOR = new AtomicLong(0);
+
+	@Override
+	public boolean supportsParameter(ParameterContext parameterContext, ExtensionContext extensionContext) {
+		if (parameterContext.isAnnotated(New.class) && parameterContext.isAnnotated(Shared.class)) {
+			String message = String
+					.format( //
+						"Parameter [%s] in %s is annotated with both @New and @Shared", //
+						parameterContext.getParameter(), testMethodDescription(extensionContext));
+			throw new ParameterResolutionException(message);
+		}
+		return parameterContext.isAnnotated(New.class) || parameterContext.isAnnotated(Shared.class);
+	}
+
+	@Override
+	public Object resolveParameter(ParameterContext parameterContext, ExtensionContext extensionContext)
+			throws ParameterResolutionException {
+		Optional<New> newAnnotation = parameterContext.findAnnotation(New.class);
+		if (newAnnotation.isPresent()) {
+			ExtensionContext.Store testStore = extensionContext.getStore(NAMESPACE);
+			Object resource = resolveNew(newAnnotation.get(), testStore);
+			return checkType(resource, parameterContext.getParameter().getType());
+		}
+
+		Optional<Shared> sharedAnnotation = parameterContext.findAnnotation(Shared.class);
+		if (sharedAnnotation.isPresent()) {
+			Parameter[] parameters = parameterContext.getDeclaringExecutable().getParameters();
+			// TODO: Replace root store with store of next outer class
+			ExtensionContext.Store rootStore = extensionContext.getRoot().getStore(NAMESPACE);
```

Comment:
Perhaps you inferred this already, but the reason why I went for the root store was to allow for use cases like, "I have a database resource that I'd like to spin up once and share with my entire test suite".

I wonder if there's a way we can keep both your suggestion and what I originally had in mind. Perhaps using a `scope` attr, as you suggested, but with all these values instead: `Class`, `SourceFile`, `Package` (if possible), and `Global`.

I think one of these should be the default. I'm tempted to go with `Global`, but that doesn't seem the safest option. `Class` and `SourceFile` are the next most intuitive for me, but `Class` has the least scope, so I'd make that the default.

Thoughts?

Reply:
@nipafx ping ^

Reply:
@nipafx I'm hesitant to move forward with the last bits of implementation and then documentation until we agree on what the API for this should be. So I was wondering what your availability to review this is like?

Reply:
Sorry for taking so long to reply. (I see, there's no shame emoji.) Lets go with `SourceFile` (default) and `Global`, initially.

I propose to leave out `Class` entirely because reusing the same name without sharing the resource in the same file sounds terribly unreadable to me. We can explore how to implement `Package` and `Module` (that would be neat) in separate PRs.